# JSON examples and exercise
****
+ get familiar with packages for dealing with JSON
+ study examples with JSON strings and files 
+ work on exercise to be completed and submitted 
****
+ reference: http://pandas.pydata.org/pandas-docs/stable/io.html#io-json-reader
+ data source: http://jsonstudio.com/resources/
****

In [1]:
import pandas as pd
import numpy as np

## imports for Python, Pandas

In [2]:
import json
from pandas.io.json import json_normalize

## JSON example, with string

+ demonstrates creation of normalized dataframes (tables) from nested json string
+ source: http://pandas.pydata.org/pandas-docs/stable/io.html#normalization

In [3]:
# define json string
data = [{'state': 'Florida', 
         'shortname': 'FL',
         'info': {'governor': 'Rick Scott'},
         'counties': [{'name': 'Dade', 'population': 12345},
                      {'name': 'Broward', 'population': 40000},
                      {'name': 'Palm Beach', 'population': 60000}]},
        {'state': 'Ohio',
         'shortname': 'OH',
         'info': {'governor': 'John Kasich'},
         'counties': [{'name': 'Summit', 'population': 1234},
                      {'name': 'Cuyahoga', 'population': 1337}]}]

In [4]:
# use normalization to create tables from nested element
json_normalize(data, 'counties')

name  population
0        Dade       12345
1     Broward       40000
2  Palm Beach       60000
3      Summit        1234
4    Cuyahoga        1337

In [5]:
# further populate tables created from nested element
json_normalize(data, 'counties', ['state', 'shortname', ['info', 'governor']])

name  population info.governor    state shortname
0        Dade       12345    Rick Scott  Florida        FL
1     Broward       40000    Rick Scott  Florida        FL
2  Palm Beach       60000    Rick Scott  Florida        FL
3      Summit        1234   John Kasich     Ohio        OH
4    Cuyahoga        1337   John Kasich     Ohio        OH

****
## JSON example, with file

+ demonstrates reading in a json file as a string and as a table
+ uses small sample file containing data about projects funded by the World Bank 
+ data source: http://jsonstudio.com/resources/

In [6]:
# load json as string
json.load((open('data/world_bank_projects_less.json')));

In [7]:
# load as Pandas dataframe
sample_json_df = pd.read_json('data/world_bank_projects_less.json')
sample_json_df

_id  approvalfy board_approval_month  \
0  {u'$oid': u'52b213b38594d8a2be17c780'}        1999             November   
1  {u'$oid': u'52b213b38594d8a2be17c781'}        2015             November   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   

  countrycode                              countryname countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia         Ethiopia   
1          TN                      Republic of Tunisia          Tunisia   

                         ...                           sectorcode source  \
0                        ...                          ET,BS,ES,EP   IBRD   
1                        ...                                BZ,BS   IBRD   

   status  supplementprojectflg  \
0  Active                     N   
1  Active                     N   

                                              theme1  \
0   {u'Percent': 100, u'Name': u'Education for all'}   
1  {u'Percent': 30, u'Name': u'Other economic man...   

                                      theme_namecode themecode   totalamt  \
0  [{u'code': u'65', u'name': u'Education for all'}]        65  130000000   
1  [{u'code': u'24', u'name': u'Other economic ma...     54,24          0   

  totalcommamt                                                url  
0    130000000  http://www.worldbank.org/projects/P129828/ethi...  
1      4700000  http://www.worldbank.org/projects/P144674?lang=en  

[2 rows x 50 columns]

****
## JSON exercises

Using data in file 'data/world_bank_projects.json' and the techniques demonstrated above,
1. Find the 10 countries with most projects
2. Find the top 10 major project themes (using column 'mjtheme_namecode')
3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

 ## Exercise 1
 First, import world bank data and store it as a dataframe named ```wb_json_df```.
 
 The output of ```wb_json_df```'s shape attribute for reveals that the dataframe contains 500 rows and 50 columns. 

In [8]:
wb_json_df = pd.read_json('data/world_bank_projects.json')
wb_json_df.head()

_id  approvalfy board_approval_month  \
0  {u'$oid': u'52b213b38594d8a2be17c780'}        1999             November   
1  {u'$oid': u'52b213b38594d8a2be17c781'}        2015             November   
2  {u'$oid': u'52b213b38594d8a2be17c782'}        2014             November   
3  {u'$oid': u'52b213b38594d8a2be17c783'}        2014              October   
4  {u'$oid': u'52b213b38594d8a2be17c784'}        2014              October   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   
2  2013-11-01T00:00:00Z   MINISTRY OF FINANCE AND ECONOMIC DEVEL   
3  2013-10-31T00:00:00Z   MIN. OF PLANNING AND INT'L COOPERATION   
4  2013-10-31T00:00:00Z                      MINISTRY OF FINANCE   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   
2                   NaN                                   Tuvalu!$!TV   
3                   NaN                        Republic of Yemen!$!RY   
4  2019-04-30T00:00:00Z                       Kingdom of Lesotho!$!LS   

  countrycode                              countryname    countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia            Ethiopia   
1          TN                      Republic of Tunisia             Tunisia   
2          TV                                   Tuvalu              Tuvalu   
3          RY                        Republic of Yemen  Yemen, Republic of   
4          LS                       Kingdom of Lesotho             Lesotho   

                         ...                           sectorcode source  \
0                        ...                          ET,BS,ES,EP   IBRD   
1                        ...                                BZ,BS   IBRD   
2                        ...                                   TI   IBRD   
3                        ...                                   JB   IBRD   
4                        ...                             FH,YW,YZ   IBRD   

   status  supplementprojectflg  \
0  Active                     N   
1  Active                     N   
2  Active                     Y   
3  Active                     N   
4  Active                     N   

                                              theme1  \
0   {u'Percent': 100, u'Name': u'Education for all'}   
1  {u'Percent': 30, u'Name': u'Other economic man...   
2  {u'Percent': 46, u'Name': u'Regional integrati...   
3  {u'Percent': 50, u'Name': u'Participation and ...   
4  {u'Percent': 30, u'Name': u'Export development...   

                                      theme_namecode    themecode   totalamt  \
0  [{u'code': u'65', u'name': u'Education for all'}]           65  130000000   
1  [{u'code': u'24', u'name': u'Other economic ma...        54,24          0   
2  [{u'code': u'47', u'name': u'Regional integrat...  52,81,25,47    6060000   
3  [{u'code': u'57', u'name': u'Participation and...        59,57          0   
4  [{u'code': u'45', u'name': u'Export developmen...        41,45   13100000   

  totalcommamt                                                url  
0    130000000  http://www.worldbank.org/projects/P129828/ethi...  
1      4700000  http://www.worldbank.org/projects/P144674?lang=en  
2      6060000  http://www.worldbank.org/projects/P145310?lang=en  
3      1500000  http://www.worldbank.org/projects/P144665?lang=en  
4     13100000  http://www.worldbank.org/projects/P144933/seco...  

[5 rows x 50 columns]

To determine the countries with the most projects, we can simply using the pandas ```.value_counts()``` method on the panda series ```wb_json_df['countryshortname']```. The result is a new pandas series (stored below as ```countrycode_counts```) that gives the number of rows containing each unique country name. Since each row correlates to a single project, this number is equal to the number of projects the World Bank currenly funds for each country. 

The ```.value_counts()``` method also sorts these counts in descending order, so to determine the 10 countries with the most projects we need only print out the first 10 elements of ```countrycode_counts```.

Doing this reveals that the 10 countries with the most World Bank funded projects are:
1. China
2. Indonesia
3. Vietnam
4. India
5. Yemen
6. Morocco
7. Bangladesh
8. Nepal
9. Africa
10. Mozambique


In [9]:
countrycode_counts = wb_json_df['countryshortname'].value_counts()
print(type(countrycode_counts))
countrycode_counts.iloc[:10]

<class 'pandas.core.series.Series'>


China                 19
Indonesia             19
Vietnam               17
India                 16
Yemen, Republic of    13
Morocco               12
Bangladesh            12
Nepal                 12
Africa                11
Mozambique            11
Name: countryshortname, dtype: int64

## Exercise 2

Before performing any analysis on the 'theme_namecode' column of the ```wb_json_df``` dataframe, we have to check to see if it contains any missing or ```NaN``` values. Using pandas' ```.info()``` method, we see that the 'theme_namecode' column has 500 non-null values. Since the ```wb_json_df``` dataframe has 500 rows, it must be the case that all the 'mjtheme_namecode' column contains no null entries. A good start!

In [10]:
wb_json_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 500 entries, 0 to 499
Data columns (total 50 columns):
_id                         500 non-null object
approvalfy                  500 non-null int64
board_approval_month        500 non-null object
boardapprovaldate           500 non-null object
borrower                    485 non-null object
closingdate                 370 non-null object
country_namecode            500 non-null object
countrycode                 500 non-null object
countryname                 500 non-null object
countryshortname            500 non-null object
docty                       446 non-null object
envassesmentcategorycode    430 non-null object
grantamt                    500 non-null int64
ibrdcommamt                 500 non-null int64
id                          500 non-null object
idacommamt                  500 non-null int64
impagency                   472 non-null object
lendinginstr                495 non-null object
lendinginstrtype            495 non

Now that we know the 'mjtheme_namecode' column only contains non-null values, we can begin to analyze it. Printing the type of the values contained in the column reveals that they are lists, and that the elements of each list are dictionaries. More specifically, each list contains a dictionary that corresponds to one of the project's themes. Each dictionary includes two keys - ```u'code'``` and ```u'name'``` - along with their values.

In [11]:
print(type(wb_json_df['mjtheme_namecode'].iloc[1]))
print(type(wb_json_df['mjtheme_namecode'].iloc[1][0]))
wb_json_df['mjtheme_namecode'].iloc[1]

<type 'list'>
<type 'dict'>


[{u'code': u'1', u'name': u'Economic management'},
 {u'code': u'6', u'name': u'Social protection and risk management'}]

Using our understanding of how the data is stored in the 'mjtheme_namecode' column, we can carry out the following steps in order to create a new dataframe that lists in the order the number of times each project theme occurs:

1. Initialize an empty dictionary to hold themes as keys and their counts as values. 
2. Loop over each row in the ```wb_json_df``` dataframe and for each row:

    - Isolate the list of dictionaries contained in the 'mjtheme_namecode' column.
    - Loop over each dictionary from the list you isolated, and for each dictionary:
        - Isolate the value associated with its 'name' key
        - If that value does not already exist as a key in your counts dictionary, set it as a new key with its value equal to 1.
        - If it already exists as a key in your counts dictionary, increment the value of that key by 1. 


 3. When finished, you should have a dictionary with all the unique theme names as keys, and how many times the theme occurs as its value. 
 4. Convert the dictionary to a dataframe.
 5. Give the dataframe appropriate column labels. 
 6. Sort the dataframe according to the values in the counts column.
 7. Reset and rename the index. 
 

In [12]:
theme_counts = {}
for index, row in wb_json_df.iterrows():
    dict_list = row['mjtheme_namecode']
    for dic in dict_list:
        theme = dic['name']
        if theme not in theme_counts.keys():
            theme_counts[theme] = 1
        else:
            theme_counts[theme] += 1

theme_counts = pd.DataFrame(theme_counts.items())
theme_counts.columns = ['Theme', 'Count']
theme_counts = theme_counts.sort_values(by=['Count'], ascending = False)
theme_counts = theme_counts.reset_index(drop = True)
theme_counts.index += 1
theme_counts.index.name = 'Place'
theme_counts.iloc[:10, :]

Theme  Count
Place                                                     
1      Environment and natural resources management    223
2                                 Rural development    202
3                                 Human development    197
4                          Public sector governance    184
5             Social protection and risk management    158
6          Financial and private sector development    130
7                                                      122
8                       Social dev/gender/inclusion    119
9                             Trade and integration     72
10                                Urban development     47

The chart above lists the 10 most common project themes. 

We can also achieve the same results by creating a list of the dictionaries from the 'mjtheme_namecode' column and using the ```json_normalize()``` function to get a dataframe containing all the theme codes and theme names for each project. The function ```make_theme_dict_list``` creates a list of all the dictionaries in a specifed dataframe column, and it is used below to create the aforementioned list of all dictionaries from the 'mjtheme_namecode' column of ```wb_json_df.``` 

Once this dataframe is created from ```json_normalize```, we can call the ```.value_counts``` method to recover the same information as in the chart above:

In [13]:
def make_theme_df(df, col):
    dict_list = []
    for index, row in df.iterrows():
        dicts = row[col]
        for dic in dicts:
            dict_list.append(dic)
    as_df = json_normalize(dict_list)
    return as_df

themes_df = make_theme_df(wb_json_df, 'mjtheme_namecode')
themes_df['name'].value_counts()

Environment and natural resources management    223
Rural development                               202
Human development                               197
Public sector governance                        184
Social protection and risk management           158
Financial and private sector development        130
                                                122
Social dev/gender/inclusion                     119
Trade and integration                            72
Urban development                                47
Economic management                              33
Rule of law                                      12
Name: name, dtype: int64

## Exercise 3

Upon re-examination of our results from question 2, we see that there is an error: the 8th most common theme is just an empty string, which means that the theme names for these codes are missing. However, we can fill in these missing theme names using the fact that we can determine the code : name pairs from the non-missing values.
   
The first step is to create a dataframe that matches each theme code with the proper theme name. To do this, we start by calling the ```make_theme_df()``` function to make a dataframe just like the one we created in question 2. Once we have this dataframe, we can run a methods chain on it so that we are left with a new dataframe that pairs each theme with its corresponding theme name. Printing this new dataframe shows that the indexes represent each code from the 'mjtheme_namecodes' column, and that each code is paired with its corresponding name.

In [14]:
codes_themes = make_theme_df(wb_json_df, 'mjtheme_namecode')
codes_themes = codes_themes.set_index('code').replace('', np.nan).dropna().drop_duplicates()
codes_themes

name
code                                              
8                                Human development
1                              Economic management
6            Social protection and risk management
5                            Trade and integration
2                         Public sector governance
11    Environment and natural resources management
7                      Social dev/gender/inclusion
4         Financial and private sector development
10                               Rural development
9                                Urban development
3                                      Rule of law

To replace the missing values in the 'mjtheme_namecodes' column, we can create a function that, for each key:value pair in the dictionaries of the column:
    
1. Determines if the name is missing.
2. If the name is missing, fill it in by matching its corresponding code using the ```codes_themes``` dictionary above.

The ```fill_values``` function below does exactly that. 

In [15]:
def fill_values(list_of_dicts):
    for dic in list_of_dicts:
            if dic[u'name'] == '':
                dic[u'name'] = codes_themes.loc[dic[u'code']][u'name']
            

In order to check that the above function works once it is applied to the ```wb_json_df``` dataframe, the cell below defines and calls the function ```print_missing_info()```, which prints out some information regarding the codes that have missing names. Since we have not yet applied the ```fill values function```, we see that there are still 122 missing names. 


In [16]:
def check_missing(df, col):
    # Count and print the total number of code:name pairs:
    num_total = df[col].value_counts().sum()
    print('Total number of code name pairs: %d' % num_total)

    # Count and print the number of codes that are missing names:
    num_missing = df[df[col] == ''][col].value_counts().sum()
    print('Number of codes that have missing names: %d' % num_missing)
    
    # Count the number of codes that are not missing names:
    num_not_missing = num_total - num_missing
    print('Number of codes that do not have missing names: %d' % num_not_missing)

    
check_missing(themes_df, 'name')

Total number of code name pairs: 1499
Number of codes that have missing names: 122
Number of codes that do not have missing names: 1377


Finally, the code in the cell below applies the fill_values function to the 'mjtheme_namecodes' column of the ```wb_json_df``` datarfame. A quick look at the first 20 elements reveals it has no missing names, a good sign!

In [17]:
wb_json_df['mjtheme_namecode'].apply(fill_values)
wb_json_df['mjtheme_namecode'].head(20)

0     [{u'code': u'8', u'name': u'Human development'...
1     [{u'code': u'1', u'name': u'Economic managemen...
2     [{u'code': u'5', u'name': u'Trade and integrat...
3     [{u'code': u'7', u'name': u'Social dev/gender/...
4     [{u'code': u'5', u'name': u'Trade and integrat...
5     [{u'code': u'6', u'name': u'Social protection ...
6     [{u'code': u'2', u'name': u'Public sector gove...
7     [{u'code': u'11', u'name': u'Environment and n...
8     [{u'code': u'10', u'name': u'Rural development...
9     [{u'code': u'2', u'name': u'Public sector gove...
10    [{u'code': u'10', u'name': u'Rural development...
11    [{u'code': u'10', u'name': u'Rural development...
12    [{u'code': u'4', u'name': u'Financial and priv...
13    [{u'code': u'5', u'name': u'Trade and integrat...
14    [{u'code': u'6', u'name': u'Social protection ...
15    [{u'code': u'10', u'name': u'Rural development...
16    [{u'code': u'10', u'name': u'Rural development...
17    [{u'code': u'8', u'name': u'Human developm

In order to make sure that ALL names were filled in, we can once again make use of the ```make_theme_df()``` and ```check_missing()``` functions. This is done in the cell below:

In [18]:
new_themes_df = make_theme_df(wb_json_df, 'mjtheme_namecode')
check_missing(new_themes_df, 'name')

Total number of code name pairs: 1499
Number of codes that have missing names: 0
Number of codes that do not have missing names: 1499


As you can see, ```wb_json_df``` has been modified so that the 'mjtheme_namecode' column no longer contains any missing names. The new value counts are also printed below, which shows that they have been updated to reflect the fact that all missing names have been filled in.

In [19]:
new_themes_df['name'].value_counts()

Environment and natural resources management    250
Rural development                               216
Human development                               210
Public sector governance                        199
Social protection and risk management           168
Financial and private sector development        146
Social dev/gender/inclusion                     130
Trade and integration                            77
Urban development                                50
Economic management                              38
Rule of law                                      15
Name: name, dtype: int64